infrastructure/src/main/java/org/corfudb/infrastructure/management/FullyConnectedClusterRecommendationEngine.java

```diff
@@ -0,0 +1,506 @@
+                                                                          .entrySet()) {
+            final String currentNode = nodeNeighborEntry.getKey();
+
+            final Map<String, Boolean> neighborsReachabilityMap =
+                    nodeNeighborEntry
+                            .getValue()
+                            .getConnectivityStatus();
+
+            // Collect nodes that are unreachable from currentNode
+            final Set<String> unreachableNeighborsSet = neighborsReachabilityMap
+                    .entrySet()
+                    .stream()
+                    .filter(neighborReachabilityEntry -> !neighborReachabilityEntry.getValue())
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toSet());
+
+            // Capture link failures seen by currentNode
+            clusterMatrix.put(currentNode, unreachableNeighborsSet);
+
+            // Capture symmetric link failures for the neighbors of currentNode
+            unreachableNeighborsSet
+                    .stream()
+                    .forEach(unreachableNeighbor ->
+                            clusterMatrix
+                                    .computeIfAbsent(unreachableNeighbor, key -> new HashSet<>())
+                                    .add(currentNode));
+
+        }
+
+        return clusterMatrix;
+    }
+
+    /**
+     * Create a map of Corfu nodes which is the superset of failed nodes along with the neighbors
+     * for each node. Each of these nodes are currently amongst the active nodes in the provided
+     * {@link Layout} and have at least one link failure with their neighboring Corfu servers.
+     *
+     * @param layout an instance of {@link Layout} representing expected layout capturing amongst
+     *               the  other things, information about responsive and unresponsive servers.
+     * @param nodeFailedNeighborMap A map which each of its entries corresponds to a Corfu
+     *                              server and a set of its neighboring nodes where a link failure
+     *                              has been observed. In other word, each entry represents a
+     *                              collection of edges from the source captured as the key of
+     *                              that entry to its failed neighbors.
+     * @return A map representing Corfu servers considered as responsive in the provided layout and
+     *         that have failed links to at least another responsive servers from the provided
+     *         layout. Mathematically the map represents a symmetric matrix which both columns
+     *         and rows represent active nodes
+     */
+    private Map<String, Set<String>> potentialFailedSuspects(Layout layout,
+                                                             Map<String, Set<String>> nodeFailedNeighborMap) {
+
+        final Map<String, Set<String>> failedLinkNodeNeighborMap = new HashMap<>();
+
+        nodeFailedNeighborMap
```

Comment:
Please add inline comments as the functional style of coding is hard to keep in context while doing code reviews. At least that is true for me.

Reply:
Done